server/vulnerabilities/cpe.go

```diff
@@ -139,67 +114,328 @@ func cleanAppName(appName string) string {
+
+	if cpeTranslationsURL == "" {
+		release, err := GetLatestNVDRelease(client)
+		if err != nil {
+			return err
+		}
+		stat, err := os.Stat(path)
+		switch {
+		case errors.Is(err, os.ErrNotExist):
+			// okay
+		case err != nil:
+			return err
+		default:
+			if stat.ModTime().After(release.CreatedAt.Time) {
+				// file is newer than release, do nothing
+				return nil
+			}
+		}
+
+		for _, asset := range release.Assets {
+			if cpeTranslationsFilename == asset.GetName() {
+				cpeTranslationsURL = asset.GetBrowserDownloadURL()
+				break
+			}
+		}
+		if cpeTranslationsURL == "" {
+			return fmt.Errorf("failed to find cpe translations in nvd release")
+
+		}
+	}
+
+	u, err := url.Parse(cpeTranslationsURL)
+	if err != nil {
+		return err
+	}
+	if err := download.DownloadAndExtract(client, u, path); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CPETranslations include special case translations for software that fail to match entries in the NVD CPE Dictionary
+// using the standard logic. This may be due to unexpected vendor or product names.
+//
+// Example:
+//
+//     [
+//       {
+//         "match": {
+//           "bundle_identifier": ["com.1password.1password"]
+//         },
+//         "translation": {
+//           "product": ["1password"],
+//           "vendor": ["agilebits"]
+//         }
+//       }
+//     ]
+type CPETranslations []CPETranslationEntry
+
+func (c CPETranslations) Translate(s *fleet.Software) (CPETranslation, bool) {
+	for _, entry := range c {
+		if entry.Match.Matches(s) {
+			return entry.Translation, true
+		}
+	}
+
+	return CPETranslation{}, false
+}
+
+type CPETranslationEntry struct {
+	Match       CPETranslationMatch `json:"match"`
+	Translation CPETranslation      `json:"translation"`
+}
+
+// CPETranslationMatch represents match criteria for cpe translations.
+type CPETranslationMatch struct {
+	Name             []string `json:"name"`
+	BundleIdentifier []string `json:"bundle_identifier"`
+	Source           []string `json:"source"`
+}
+
+func (c CPETranslationMatch) Matches(s *fleet.Software) bool {
+	if len(c.Name) > 0 {
+		found := false
+		for _, name := range c.Name {
+			if name == s.Name {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return false
+		}
+	}
+	if len(c.BundleIdentifier) > 0 {
+		found := false
+		for _, bundleID := range c.BundleIdentifier {
+			if bundleID == s.BundleIdentifier {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return false
+		}
+	}
+	if len(c.Source) > 0 {
+		found := false
+		for _, source := range c.Source {
+			if source == s.Source {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return false
+		}
+	}
+	return true
+}
+
+type CPETranslation struct {
+	Product  []string `json:"product"`
+	Vendor   []string `json:"vendor"`
+	TargetSW []string `json:"target_sw"`
+}
+
+func CPEFromSoftware(db *sqlx.DB, software *fleet.Software, translations CPETranslations) (string, error) {
```

Comment:
While we are at it, please document how CPEFromSoftware does the matching (first it tries translations, otherwise resorts to NVD post-processed sqlite3, etc.).

Reply:
Added some documentation.